### DISCUSSION TITLE
Enable use of Github App installation tokens instead of PATS

### DISCUSSION THREAD

**Discussion by lenmazzone** [upptime/upptime#793] [Ideas] [answered]
Hi,

My organization would like to use Upptime with a Github App instead of having a service account just to issue a PAT. Is this something you could support? 

See: 
https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/making-authenticated-api-requests-with-a-github-app-in-a-github-actions-workflow

https://github.com/marketplace/actions/workflow-application-token-action

**Comment by AnandChowdhary** [comment, marked as answer]
This is not currently planned because it requires setting up a backend, and Upptime currently works directly on GitHub, no external server needed.

**Reply by Andre601** [reply]
I don't think it does require a backend...

You create an app, generate the private key and save it as Actions secret and then using the `github-app-token` action to generate a PAT to use (which afaik is temporary, so it would be discarded after a while).

There is no step that would require you to have a backend to store any information.
As a proof, here is an action workflow I use for automated issue/PR comments using a App rather than an account: https://github.com/Readme-Workflows/recent-activity/blob/main/.github/workflows/comment.yml

If you mean that every workflow would now require a extra job to determine if an app is used rather than a user account, then yeah, that could be problematic in a way...
Tho I also think it wouldn't be too bad either? Didn't github have a way to define commonly used workflows/steps/jobs?

Either way, the easiest way I can imagine this to work is having two jobs with oposite if-conditions checking for if a `APP_ID` ENV is defined. If it is, run the job that uses the app-token action to get a PAT to use, if not default to a PAT.

Speaking of, looking at the latest workflows I noticed the usage of `${{ secrets.GH_PAT || github.token }}`. Does it mean the action now properly supports the usage of the Actions user rather than a custom one, meaning I can discontinue having a dedicated account just for this?

**Reply by AnandChowdhary** [reply]
You're right, I missed that! Definitely sounds interesting to explore.

Yes, with the right permissions (they are in the docs IIRC) you should be able to use the default Actions user.

**Reply by yahel2410** [reply]
Hi @AnandChowdhary, our org would also love to see GH app support. Would you like me to make a PR with the changes?

**Reply by AnandChowdhary** [reply]
Yes, happy to accept a PR!

**Reply by yahel2410** [reply]
I see that the `GH_PAT` is going to be removed here https://github.com/upptime/upptime/pull/781
Should I exclude it from the PR?

**Reply by Andre601** [reply]
Given how old the last activity there is, I think you'll be fine toi ignore it for now...

**Reply by AnandChowdhary** [reply]
Yeah no worries with keeping it for now :)

**Comment by Jackenmen** [comment]
GitHub now offers an official action to create an app token using its client id and private key:
https://github.com/actions/create-github-app-token
Mentioned in the docs here:
https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/making-authenticated-api-requests-with-a-github-app-in-a-github-actions-workflow

Any chance this could be reconsidered? Machine accounts can be a bit annoying to manage in an org setting compared to GH Apps (and generating a *personal* org-scoped fine-grained token means that the actor is the user that generated the token).

**Comment by martijnvankekem** [comment]
@AnandChowdhary @Jackenmen 

I created this discussion precisely indicating what would be necessary to implement this currently.

https://github.com/orgs/upptime/discussions/1173
